### ISSUE TITLE
Pattern: Using commit hooks as a mean to communicate between apps

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3839]
We recently discussed a pattern in use that utilize [commit hooks](http://starcounter.io/guides/transactions/commit-hooks/) as a mean to notify / communicate between apps.

**Shared.dll**
```cs
namespace Shared {
  [Database] public class Foo {}
}
```

**App1.exe**
```cs
using Shared;
class Program {
  void Main() {
    Db.Transact(() => new Foo());
  }
}
```

**App2.exe**
```cs
using Shared;
[Database] public class FooInserted {}
class Program {
  void Main() {
    Hook<Foo>.CommitInsert += (s, obj) => {
      Db.Transact(() => new FooInserted());
    };
  }
}
```

Is this one of our recommended patterns, one we want to promote? Are there alternatives?

cc @warpech @miyconst @chrhol @k-rus

**Comment by warpech** [comment]
Yes, this is a pattern that we are using right now. Personally I think this is the only possible pattern to integrate apps that don't know about each other.

Note: As @aloscha reports, right now you cannot use `Db.Transact` in a commit hook, because commit hook is already in a transaction. There is a pattern to bypass that: https://github.com/Starcounter/Starcounter/issues/2766.

Code that shows this in action:  https://github.com/Starcounter/WarehouseClient/blob/5bf73cd810b80ea7df7795f6624cfecdbf804978/src/Program.cs#L56

**Comment by per-samuelsson** [comment]
Thanks for giving feedback, @warpech.

> Note: As @aloscha reports, right now you cannot use Db.Transact in a commit hook, because commit hook is already in a transaction.

We need to straighten this out, because it sound confusing. If hooks are used for this kind of thing, we really need to document them more properly IMO.

What exactly do you mean that `Db.Transact` **cannot be used** in the context of a hook? What exactly happen? Operations in the scope of `Db.Transact` just get part of the outer transaction, no? Isn't this the same thing as doing:

```cs
void Foo() {
  Db.Transact(() => new Foo());
}

void Bar() {
  Db.Transact(() => Foo());
}
```

cc @chrhol

**Comment by k-rus** [comment]
I think it will be necessary to design a mechanism to share database changes (not necessary commits) between apps on common/shared data. However, no thoughts were put in this. So it is future, not something should be taken in account now. For now I don't have any good suggestion for the pattern.

**Comment by chrhol** [comment]
> What exactly do you mean that Db.Transact cannot be used in the context of a hook? What exactly happen?

I guess the problem is long-running (fuzzy) transactions, not `Db.Transact`. You cannot switch from one to another inside a hook, and have to schedule a task instead.

**Comment by per-samuelsson** [comment]
> I think it will be necessary to design a mechanism to share database changes (not necessary commits) between apps on common/shared data.

I do too.

To start with, we should document very precisely the do's and dont's regarding this, if it is like @warpech says, that this approach is the one we recommend for app developers.

One of these is that things can go very wrong if applications use this technique to ensure things like "for every order that is ever placed, I want to be sure informing the IT-department about it":

```cs
Hook<Order>.CommitInsert += (s, obj) => { PassItOn(obj); }
```

The main problem with it is that for orders inserted *while the "hooking app" is not running*, **no notifications** will ever be sent.

**Comment by warpech** [comment]
FYI we have a new [Cookbook](http://starcounter.io/cookbook/) section on .io that should evolve into a presentation of patterns and other "how do I..." material.

**Comment by per-samuelsson** [comment]
> FYI we have a new [Cookbook](http://starcounter.io/cookbook/) section on .io that should evolve into a presentation of patterns and other "how do I..." material.

That's a great initiative. Rename it "How to"? Think that is more standard for that kind of thing, no?

As for hook documentation specifically, we should expand the [dedicated page](http://starcounter.io/guides/transactions/commit-hooks/) IMO. And add sections covering:

1. That hooked types don't expand to subtypes of said type (common pitfall).
2. That hooks are not invoked for CRUD ops executed in the scope of a hook.
3. That hooks execute only on real-time events, meaning that operations you would receive a call for that has occurred while your application was not running is not triggering them.
